Port np.roots

mathisgerdes opened the pull request:
Implements the numpy roots function (c.f. #70).

At least on my CPU, this version seems a bit slower than the numpy version. I think the main reasons are that slice indexing and the eigvals functions are a bit slower, this version is as fast as I could get it.

Because one has to remove leading zeros from the input array, the full np.roots function can't be jitted. I didn't find a way around this.

mathisgerdes committed:
Implement np.roots.

mathisgerdes committed:
Merge branch 'master' of github.com:google/jax

shoyer commented:
To make this `jit` compatible, we need to get rid of the trimming step so that all the arrays have a fixed shape.

As I understand it, there are two reasons why NumPy does trimming:
1. Leading zeros are trimmed because the [companion matrix](https://en.wikipedia.org/wiki/Companion_matrix) approach to computing eigenvalues requires that coefficient of the highest order polynomial is 1 (which would otherwise entail dividing by zero).
2. Trailing zeros are trimmed as an optimization, because we know they simply add additional roots at zero.

So I think we can avoid this trimming and make this algorithm `jit` compatible by skipping the optimization from (2) and by replacing the trimming in (1) with `np.roll` to move leading zeros to the end (effectively multiplying by an integer power of `x`, which doesn't change the roots). We'll also need a check for the all zero coefficient case, to avoid dividing by zero, which would introduce a NaN entry in the companion matrix.

What do you think?

mathisgerdes commented:
I think that approach almost works, the trailing zeros are, as you say, just an optimization in the algorithm. The fundamental problem that still exists as far as I can see is that the number of roots depends on the number of leading zeros (since the order of the polynomial isn't given by the input length in that case). This inevitably leads to a value dependent output shape and thus problems with jit.

In my own application where I use this, I can guarantee leading zeros are stripped which makes it much faster (and can jit). Maybe one could expose a fast version next to the safe version?

shoyer commented:
> I think that approach almost works, the trailing zeros are, as you say, just an optimization in the algorithm. The fundamental problem that still exists as far as I can see is that the number of roots depends on the number of leading zeros (since the order of the polynomial isn't given by the input length in that case). This inevitably leads to a value dependent output shape and thus problems with jit. Your version would yield the wrong multiplicity of zero roots.

Good catch, this is too bad!

> In my own application where I use this, I can guarantee leading zeros are stripped which makes it much faster (and can jit). Maybe one could expose a fast version next to the safe version?

Yes, sounds like a good idea! I would suggest implementing this via a new, keyword-only argument of some form. The default should have the dynamic shape like `np.roots` but you could opt into another version that is `jit` compatible.

mathisgerdes committed:
Expose jit-compatible variant of np.roots.

General np.roots implementation has a value dependent output shape.
If the input coefficients are guaranteed to have no leading zeros,
output shape is independent of values. Skip checking for leading
zeros by setting a keyword argument.

mathisgerdes commented:
I've added a keyword to np.roots so can now apply jit to one variant: 

```python
>>> roots_unsafe = jax.jit(jax.partial(np.roots, strip_zeros=False))
>>> roots_unsafe([1, 2])     # ok
DeviceArray([-2.+0.j], dtype=complex64)
>>> roots_unsafe([0, 1, 2])  # problem
DeviceArray([nan+nanj, nan+nanj], dtype=complex64)
>>> roots([0, 1, 2])         # use the no-jit version instead
DeviceArray([-2.+0.j], dtype=complex64)
```

mathisgerdes committed:
Merge branch 'master' of github.com:google/jax

mathisgerdes committed:
Fix typo.

shoyer commented on jax/numpy/polynomial.py:
```diff
@@ -0,0 +1,104 @@
+>>> np.roots([0, 1, 2])         # use the no-jit version instead
+DeviceArray([-2.+0.j], dtype=complex64)
+""")
+def roots(p, strip_zeros=True):
```
Just to be extra cautious, could you make this keyword only?
```suggestion
def roots(p, *, strip_zeros=True):
```

shoyer commented on jax/numpy/polynomial.py:
```diff
@@ -0,0 +1,104 @@
+
+The general implementation can therefore not be transformed with jit.
+If the coefficients are guaranteed to have no leading zeros, use the 
+keyword argument `strip_zeros=False` to get a jit-compatible variant:
+
+>>> roots_unsafe = jax.jit(jax.partial(np.roots, strip_zeros=False))
+>>> roots_unsafe([1, 2])     # ok
+DeviceArray([-2.+0.j], dtype=complex64)
+>>> roots_unsafe([0, 1, 2])  # problem
+DeviceArray([nan+nanj, nan+nanj], dtype=complex64)
+>>> np.roots([0, 1, 2])         # use the no-jit version instead
+DeviceArray([-2.+0.j], dtype=complex64)
```
If you use `::` and an indent, I think restructured text will parse this as code block in the online docs:
```suggestion
keyword argument `strip_zeros=False` to get a jit-compatible variant::

    >>> roots_unsafe = jax.jit(jax.partial(np.roots, strip_zeros=False))
    >>> roots_unsafe([1, 2])     # ok
    DeviceArray([-2.+0.j], dtype=complex64)
    >>> roots_unsafe([0, 1, 2])  # problem
    DeviceArray([nan+nanj, nan+nanj], dtype=complex64)
    >>> np.roots([0, 1, 2])         # use the no-jit version instead
    DeviceArray([-2.+0.j], dtype=complex64)
```

shoyer commented on tests/polynomial_test.py:
```diff
@@ -0,0 +1,61 @@
+    np_fn = lambda arg: onp.sort(np.roots(arg))
+    onp_fn = lambda arg: onp.sort(onp.roots(arg))
+    self._CheckAgainstNumpy(onp_fn, np_fn, args_maker, check_dtypes=False)
+
```
Could you please add a test that exercises `strip_zeros=False`?

Probably would be good to check:
- Something valid
- Something with a leading zero
- Size 0 input

mathisgerdes replied:
For the "something with a leading zero" test, do you mean to test that the output array indeed contains np.nan's?

shoyer replied:
Yes, exactly that!

shoyer reviewed (approved):
Sorry for the delay getting back to you here -- this looks great!

I have only a few minor suggestions/requests.

To also add this to the online docs, could you add it to the list in `docs/jax.numpy.rst`?

mathisgerdes committed:
Make roots jit-argument keyword only.

Co-Authored-By: Stephan Hoyer <shoyer@google.com>

mathisgerdes committed:
Format docstring to enable parsing.

Co-Authored-By: Stephan Hoyer <shoyer@google.com>

mathisgerdes committed:
Add np.roots function to documentation.

mathisgerdes committed:
Merge branch 'master' of github.com:google/jax

mathisgerdes committed:
Add more tests for np.roots function.

- Include length 0 polynomial coefficients
- Test strip_zeros=False argument
- Test jit compiled version (only on cpu due to eigvals)
- Confirm that adding leading zeros while skipping check
  for them results in nan's (expected behavior)

mathisgerdes commented:
Thank you for the suggestions! 

> Probably would be good to check:
> 
> * Something valid
> * Something with a leading zero
> * Size 0 input

I have added tests that include size 0 inputs and setting `strip_zeros=False`.
For `strip_zeros=False` and coefficients with leading zeros the output array is checked to contain nan's. The alternative would be for np.roots to somehow raise an exception if there are leading zeros despite `strip_zeros=False` (I haven't figured out how to make that work together with jit, if it is possible). Given that at the moment nan's are the expected behavior, I suppose the test at least makes that clear.

mathisgerdes committed:
Fix bug in np.roots test.

The polynomial with coefficents [0] never fails because the number of
roots is 0.

mathisgerdes committed:
Avoid bug in eigvals and adjust test accuracy.

The parameters of the test that was changed are non-essential
since they test for how the code behaves given invalid inputs.

The accuracy in comparing to the numpy result is changed because
the algorithm in those cases is slightly changed with respect to
the original numpy algorithm (to allow jit).

mathisgerdes commented:
The latest test fails, I believe, because of a bug in `np.eigvals`. I opened an issue describing it here #2661.

shoyer commented:
Thanks @mathisgerdes!
